### ISSUE TITLE
Limitation on size of received data

### ISSUE THREAD

**Issue by mdbartos** [open-storm/perfect-cell#31]
Influxdb sends chunked responses back to the node. However, the current implementation checks for a maximum of two chunks. This should be extended to an arbitrary number of chunks (so long as the total size is less than the available receive buffer size).

**Comment by mdbartos** [comment]
Here's a Python mockup of the solution to this issue. Just need to check for bugs and translate to C:

```
# Acquire IP address
send_at_command("AT#SGACT=1,1",10)
# Socket dial
send_at_command('AT#SD=1,0,8086,"data.open-storm.org",0,0,1',10)
# Open send context
send_at_command("AT#SSEND=1")

# Download a random number of vbat measurements from 1-100
send_at_command(("GET "
                 "/query?db=REDACTED&u=REDACTED&p=REDACTED&q=SELECT%20value%20from%20v_bat%20where%20node_id='GGB006'%20LIMIT%20{0} "
                  "HTTP/1.1\r\nHost: data.open-storm.org:8086\r\nConnection: "
                  "Close\r\n\r\n\032").format(random.randint(1, 100)), r=False)

# Make sure read buffer is empty
sleep(1)
send_at_command('')

def get_chunked_response(response, max_loops=100):
    message = ''
    buffer_start = 1
    chunk_start = 1
    buffer_num = 0
    # In the C version, b will need to be a pointer to the beginning of the
    # response string
    a = 0
    b = 0

    for loop in range(max_loops):
        error = response.find('CMEE ERROR')
        message_end = response.find('\r\n\r\nOK\r\n')
        ending_buffer = response.find('\r\n0\r\n')

        if (error != -1) or (message_end == -1):
            print('CMEE ERROR')
            return 0

        # Need to check for null pointers on all of these
        a = response.find('\r\n', b) + len('\r\n')
        if buffer_start:
            a = response.find('\r\n', a) + len('\r\n')
            buffer_start = 0
        b = response.find('\r\n', a)

        if chunk_start:
            chunk_size = int(response[a:b], 16)
            chunk_remainder = chunk_size
            a = b + len('\r\n')
            b = response.find('\r\n', a)
            chunk_start = 0

        # Remember to add check for message buffer size
        message += response[a:b]
        remainder_len = b - a
        chunk_remainder -= remainder_len

        if not chunk_size:
            return message

        if b >= message_end:
            if b > message_end:
                print('End pointer beyond message end')
                return 0
            sleep(1)
            response = send_at_command("AT#SRECV=1,1500").decode('ascii')
            buffer_num += 1
            buffer_start = 1
            a = 0
            b = 0

        if chunk_remainder <= 0:
            if chunk_remainder < 0:
                print('Invalid chunk remainder')
                return 0
            if chunk_size != 0:
                chunk_start = 1
    return 0

# Get first response
response = send_at_command("AT#SRECV=1,1500").decode('ascii')

# Look for SRING
sring = response.find('SRING: 1')

if (sring != -1):
    print('SRING')
    sleep(1)
    response = send_at_command("AT#SRECV=1,1500").decode('ascii')
    message = get_chunked_response(response)
else:
    print('NO SRING')
    sleep(1)
    start = response.find('Transfer-Encoding: chunked')
    response = response[start:]
    message = get_chunked_response(response)

send_at_command('AT#SH=1')
```

**Comment by mdbartos** [comment]
C implementation. It seems to work, but I need to test it on some more examples:

```
int read_chunked_response(char message[], char *response_start, 
                          int max_loops, int max_message_size){

    int message_free_space = max_message_size - strlen(message);
    int chunk_remainder = 0;
    int remainder_len = 0;
    int buffer_start = 1;
    int chunk_start = 1;
    int buffer_num = 0;
    int chunk_size = 0;
    int status = 0;
    char *error = NULL;
    char *message_end = NULL;
    char *ending_buffer = NULL;
    char *a = response_start;
    char *b = response_start;
    char hexstring[10] = {'\0'};
    int iter;

    for (iter=0; iter < max_loops; iter++){
        // Check for CMEE error
        error = strstr(response_start, "CMEE ERROR");
        // Get pointer to end of Telit buffer
        message_end = strstr(response_start, "\r\n\r\nOK\r\n");
        // Check if this buffer contains the terminating chunk and get pointer
        ending_buffer = strstr(response_start, "\r\n0\r\n");

        // Break if CMEE error or message truncated
        if (error || !message_end){
            return 0u;
        }

        // Move the beginning pointer to the start of the next line
        a = strstr(b, "\r\n") + strlen("\r\n");
        if (!a) {return 0u;}
        // If we are at the beginning of the Telit buffer, skip another line
        if (buffer_start){
            a = strstr(a, "\r\n") + strlen("\r\n");
            if (!a) {return 0u;}
            // Reset buffer start flag
            buffer_start = 0;
        }
        // Move the end pointer to the end of the line
        b = strstr(a, "\r\n");
        if (!b) {return 0u;}

        // If we're positioned at the start of a new chunk, get the chunk size
        if (chunk_start){
            // Get new chunk size
            strncpy(hexstring, a, b-a);
            chunk_size = (int)strtol(hexstring, NULL, 16);
            memset(hexstring, '\0', sizeof(hexstring));
            chunk_remainder = chunk_size;
            // Move the beginning pointer to the start of the next line
            a = b + strlen("\r\n");
            if (!a) {return 0u;}
            // Move the end pointer to the end of the line
            b = strstr(a, "\r\n");
            if (!b) {return 0u;}
            // Reset chunk start flag
            chunk_start = 0;
        }

        memset(current_selection, '\0', sizeof(current_selection));
        strncpy(current_selection, a, b - a);

        // Add the current selection to the message buffer
        // First, check for buffer overflow
        message_free_space = max_message_size - strlen(message);
        if ((b - a) >= message_free_space){
            strncat(message, a, message_free_space);
            return 0u;
        }
        // Concatenate the latest selection to the message
        strncat(message, a, b - a);
        // Compute the remaining bytes in the chunk
        chunk_remainder -= (b - a);

        // If we've reached the terminal chunk, exit the function
        if (!chunk_size){
            return 1u;
        }

        // Check if current Telit buffer is exhausted
        if (b >= message_end){
            // Make sure that we haven't gone over the end of the buffer
            if (b > message_end){
                return 0u;
            }
        // If exhausted, get the next Telit buffer
        buffer_num += 1;
        // This is a fake version of the at_write_command function; don't worry about it ;-)
        status = at_write_command(recv_cmd, "OK", buffer_num);
        // Set the flag to indicate we're at the start of a Telit buffer
        buffer_start = 1;
        // Reset pointers
        response_start = uart_received_string;
        a = response_start;
        b = response_start;
        }

        // Check chunk status
        if (chunk_remainder <= 0){
            // Check to see if we've overflowed the end of the chunk
            if (chunk_remainder < 0){
                return 0u;
            }
            // If chunk is exhausted, set flag to indicate we're at the 
            // start of a new chunk
            if (chunk_size != 0){
                chunk_start = 1;
            }
        }
    }
    return 0u;
}
```

**Comment by mdbartos** [comment]
Addressed in #41
